sdk/storage/azure-storage-files-datalake/test/ut/datalake_sas_test.cpp

```diff
@@ -207,271 +327,480 @@ namespace Azure { namespace Storage { namespace Test {
+          std::make_shared<Azure::Identity::ClientSecretCredential>(
+              AadTenantId(), AadClientId(), AadClientSecret(), GetTokenCredentialOptions()),
+          InitStorageClientOptions<Files::DataLake::DataLakeClientOptions>());
+      userDelegationKey = dataLakeServiceClient.GetUserDelegationKey(sasExpiresOn).Value;
     }
 
-    // Encryption scope
+    std::string fileName = RandomString();
+
+    auto dataLakeFileSystemClient = *m_fileSystemClient;
+    auto dataLakeFileClient = dataLakeFileSystemClient.GetFileClient(fileName);
+    dataLakeFileClient.Create();
+
+    Sas::DataLakeSasBuilder fileSasBuilder;
+    fileSasBuilder.Protocol = Sas::SasProtocol::HttpsAndHttp;
+    fileSasBuilder.ExpiresOn = sasExpiresOn;
+    fileSasBuilder.Resource = Sas::DataLakeSasResource::File;
+    fileSasBuilder.FileSystemName = m_fileSystemName;
+    fileSasBuilder.Path = fileName;
+    fileSasBuilder.SetPermissions(Sas::DataLakeSasPermissions::All);
+    fileSasBuilder.PreauthorizedAgentObjectId = RandomUUID();
+    fileSasBuilder.CorrelationId = RandomUUID();
+    auto sasToken = fileSasBuilder.GenerateSasToken(userDelegationKey, accountName);
+    VerifyDataLakeSasRead(dataLakeFileClient, sasToken);
+  }
+
+  TEST_F(DataLakeSasTest, FileSasWithIdentifier)
+  {
+    auto sasStartsOn = std::chrono::system_clock::now() - std::chrono::minutes(5);
+    auto sasExpiresOn = std::chrono::system_clock::now() + std::chrono::minutes(60);
+
+    auto keyCredential
+        = _internal::ParseConnectionString(StandardStorageConnectionString()).KeyCredential;
+
+    std::string fileName = RandomString();
+
+    auto dataLakeFileSystemClient = *m_fileSystemClient;
+    auto dataLakeFileClient = dataLakeFileSystemClient.GetFileClient(fileName);
+    dataLakeFileClient.Create();
+
+    Files::DataLake::SetFileSystemAccessPolicyOptions options;
+    options.AccessType = Files::DataLake::Models::PublicAccessType::None;
+    Files::DataLake::Models::SignedIdentifier identifier;
+    identifier.Id = RandomString(64);
+    identifier.StartsOn = sasStartsOn;
+    identifier.ExpiresOn = sasExpiresOn;
+    identifier.Permissions = "r";
+    options.SignedIdentifiers.emplace_back(identifier);
+    dataLakeFileSystemClient.SetAccessPolicy(options);
+
+    Sas::DataLakeSasBuilder fileSasBuilder;
+    fileSasBuilder.Protocol = Sas::SasProtocol::HttpsAndHttp;
+    fileSasBuilder.ExpiresOn = sasExpiresOn;
+    fileSasBuilder.FileSystemName = m_fileSystemName;
+    fileSasBuilder.Path = fileName;
+    fileSasBuilder.Resource = Sas::DataLakeSasResource::File;
+    fileSasBuilder.SetPermissions(static_cast<Sas::DataLakeFileSystemSasPermissions>(0));
+    fileSasBuilder.Identifier = identifier.Id;
+
+    // TestSleep(std::chrono::seconds(30));
+
+    auto sasToken = fileSasBuilder.GenerateSasToken(*keyCredential);
+
+    // TODO: looks like a server bug, the identifier doesn't work sometimes.
+    // VerifyDataLakeSasRead(dataLakeFileClient, sasToken);
```

Comment:
Can you check BlobSasTest.BlobSasWithIdentifier? This issue seems to have been fixed

Reply:
Fixed. It's because I use the blob connection string instead of datalake connection stirng by mistake.

Another thing is that why we should sleep 30 seconds? To wait for the policy? Can we sleep less, such as 10 seconds? This may reduce the time for test.